### ISSUE TITLE
[MNG-7621] Parameter '-f' causes ignoring any 'maven.config' (only on Windows)

### ISSUE THREAD

**Issue by jira-importer** [apache/maven#8832]
**[Oliver Glowa](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=ollily)** opened **[MNG-7621](https://issues.apache.org/jira/browse/MNG-7621?redirect=false)** and commented

Hello,

first of all, I didn't find a ticket, related to the possible bug I found (keyword "-f" and "maven.config")
https://issues.apache.org/jira/issues/?jql=project%20%3D%20MNG%20AND%20text%20~%20maven.config%20AND%20text%20~%20-f

# Analysis

Here the problem I found:

When executing maven with parameter "-f" without a file name (e.g. "pom.xml") any ".mvn\maven.config" is ignored.

```
C:\projects\projects_github>mvn -f maven-test clean verify
[INFO] Scanning for projects...
[INFO]
[INFO] --------------------< com.glowanet.appl:maven-test >--------------------
[INFO] Building maven-test 1.0-SNAPSHOT
[INFO] --------------------------------[ jar ]---------------------------------
[INFO]
[INFO] --- maven-clean-plugin:2.5:clean (default-clean) @ maven-test ---
[INFO] Deleting C:\projects\projects_github\maven-test\target
[INFO]
[INFO] --- echo-maven-plugin:1.3.2:echo (echo-plugin) @ maven-test ---
[INFO] plugin active
[INFO]
[INFO] --- maven-resources-plugin:2.6:resources (default-resources) @ maven-test ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] Copying 0 resource
[INFO]
[INFO] --- maven-compiler-plugin:3.1:compile (default-compile) @ maven-test ---
[INFO] Changes detected - recompiling the module!
[INFO] Compiling 1 source file to C:\projects\projects_github\maven-test\target\classes
[INFO]
[INFO] --- maven-resources-plugin:2.6:testResources (default-testResources) @ maven-test ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] skip non existing resourceDirectory C:\projects\projects_github\maven-test\src\test\resources
[INFO]
[INFO] --- maven-compiler-plugin:3.1:testCompile (default-testCompile) @ maven-test ---
[INFO] Nothing to compile - all classes are up to date
[INFO]
[INFO] --- maven-surefire-plugin:2.12.4:test (default-test) @ maven-test ---
[INFO] No tests to run.
[INFO]
[INFO] --- maven-jar-plugin:2.4:jar (default-jar) @ maven-test ---
[INFO] Building jar: C:\projects\projects_github\maven-test\target\maven-test-1.0-SNAPSHOT.jar
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  2.063 s
[INFO] Finished at: 2022-12-07T14:21:45+01:00
[INFO] ------------------------------------------------------------------------

```

 

When the filename is added, the "maven.config" will be recognized (the parameter in "maven.config" will activated the profile "profile1")

```
C:\projects\projects_github>mvn -f maven-test\pom.xml clean verify
[INFO] Scanning for projects...
[INFO]
[INFO] --------------------< com.glowanet.appl:maven-test >--------------------
[INFO] Building maven-test 1.0-SNAPSHOT
[INFO] --------------------------------[ jar ]---------------------------------
[INFO]
[INFO] --- maven-clean-plugin:2.5:clean (default-clean) @ maven-test ---
[INFO] Deleting C:\projects\projects_github\maven-test\target
[INFO]
[INFO] --- maven-help-plugin:3.3.0:active-profiles (default) @ maven-test ---
[INFO]
Active Profiles for Project 'com.glowanet.appl:maven-test:jar:1.0-SNAPSHOT':

The following profiles are active: 

- profile1 (source: com.glowanet.appl:maven-test:1.0-SNAPSHOT)



[INFO]
[INFO] --- echo-maven-plugin:1.3.2:echo (echo-plugin) @ maven-test ---
[INFO] plugin active
[INFO]
[INFO] --- echo-maven-plugin:1.3.2:echo (echo-profile) @ maven-test ---
[INFO] profile1 active
[INFO]
[INFO] --- maven-resources-plugin:2.6:resources (default-resources) @ maven-test ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] Copying 0 resource
[INFO]
[INFO] --- maven-compiler-plugin:3.1:compile (default-compile) @ maven-test ---
[INFO] Changes detected - recompiling the module!
[INFO] Compiling 1 source file to C:\projects\projects_github\maven-test\target\classes
[INFO]
[INFO] --- maven-resources-plugin:2.6:testResources (default-testResources) @ maven-test ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] skip non existing resourceDirectory C:\projects\projects_github\maven-test\src\test\resources
[INFO]
[INFO] --- maven-compiler-plugin:3.1:testCompile (default-testCompile) @ maven-test ---
[INFO] Nothing to compile - all classes are up to date
[INFO]
[INFO] --- maven-surefire-plugin:2.12.4:test (default-test) @ maven-test ---
[INFO] No tests to run.
[INFO]
[INFO] --- maven-jar-plugin:2.4:jar (default-jar) @ maven-test ---
[INFO] Building jar: C:\projects\projects_github\maven-test\target\maven-test-1.0-SNAPSHOT.jar
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  2.123 s
[INFO] Finished at: 2022-12-07T14:22:26+01:00
[INFO] ------------------------------------------------------------------------
```

## Additional

The problem occurs only on a windows commandline, not under linux (cygwin) shell.

## Test with

* Maven 3.6.3 / Maven 3.8.6
* OpenJDK 17
* Windows 10 64Bit
* Cygwin 64Bit

```
Apache Maven 3.6.3 (cecedd343002696d0abb50b32b541b8a6ba2883f)
Maven home: C:\Programme-2\apache-maven\apache-maven\bin\..
Java version: 17.0.4.1, vendor: Eclipse Adoptium, runtime: C:\Programme-2\openjdk\jdk
Default locale: de_DE, platform encoding: Cp1252
OS name: "windows 10", version: "10.0", arch: "amd64", family: "windows"

Apache Maven 3.8.6 (84538c9988a25aec085021c365c560670ad80f63)
Maven home: C:\Programme-2\apache-maven\apache-maven-3.8.6
Java version: 17.0.4.1, vendor: Eclipse Adoptium, runtime: C:\Programme-2\openjdk\jdk
Default locale: de_DE, platform encoding: Cp1252
OS name: "windows 10", version: "10.0", arch: "amd64", family: "windows"
```

## Example

An example project is added .

```
https://github.com/ollily/maven-test.git
```

call "build.bat" to see the different handling.

 

---

**Affects:** 3.6.3, 3.8.6

**Issue Links:**
- [MNG-8056](https://issues.apache.org/jira/browse/MNG-8056) Parameter '-f' & '-s' fails with absolute paths on Cygwin

**Remote Links:**
- [GitHub Pull Request #905
](https://github.com/apache/maven/pull/905)

**Backported to:** [4.0.0-alpha-4](https://github.com/apache/maven/milestone/3?closed=1), [3.9.0](https://github.com/apache/maven/milestone/2?closed=1), [3.8.7](https://github.com/apache/maven/milestone/108?closed=1)

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

[sjaranowski](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=sjaranowski), this requires an IT because I am quite surprised how we missed this.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

I think I have found the cause and a tentative fix.
Block in question:

```
:findBaseDir
cd /d "%WDIR%"
:findBaseDirLoop
if exist "%WDIR%\.mvn" goto baseDirFound
cd ..
IF "%WDIR%"=="%CD%" goto baseDirNotFound
set "WDIR=%CD%"
goto findBaseDirLoop
```

Now `%WDIR%` is a relative directory in your case and since we change into it the `if exist` does a test which is nested too deep: `if exist maven-test\.mvn` **in** `maven-test` which does not make any sense. If we make it to `if exist ".mvn" goto baseDirFound` it still would be wrong because the next test compares are relative with an absolute directory which will always fail in your case. Sample: `C:\Entwicklung\Projekte\maven-test>IF "maven-test\target" == "C:\Entwicklung\Projekte\maven-test" goto baseDirNotFound`

In my opinion there is one solution which needs to be tested properly:

```
:findBaseDir
cd /d "%WDIR%"
set "WDIR=%CD%"
:findBaseDirLoop
if exist ".mvn" goto baseDirFound
cd ..
IF "%WDIR%"=="%CD%" goto baseDirNotFound
set "WDIR=%CD%"
goto findBaseDirLoop
```

Here we make sure that after the directory is changed that `WDIR` is always absolute.

[Oliver Glowa](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=ollily), can you try that single line change on your end? For me, this would qualify for 3.8.7.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

My reference: https://stackoverflow.com/a/4488734/696632

**Comment by jira-importer** [comment]
**[Oliver Glowa](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=ollily)** commented

Ok,
what have I done:

Opened <MAVEN_HOME>\bin\mvn.cmd (Release 3.8.6)
Edit line 140

```
OLD:    if exist "%WDIR%\.mvn" goto baseDirFound
NEW:    if exist ".mvn" goto baseDirFound
```

Test with my example (https://github.com/ollily/maven-test.git)

```
C:\projects\projects_github>mvn -f maven-test clean verify
[INFO] Scanning for projects...
[INFO]
[INFO] --------------------< com.glowanet.appl:maven-test >--------------------
[INFO] Building maven-test 1.0-SNAPSHOT
[INFO] --------------------------------[ jar ]---------------------------------
[INFO]
[INFO] --- maven-clean-plugin:2.5:clean (default-clean) @ maven-test ---
[INFO] Deleting C:\projects\projects_github\maven-test\target
[INFO]
[INFO] --- maven-help-plugin:3.3.0:active-profiles (default) @ maven-test ---
[INFO]
Active Profiles for Project 'com.glowanet.appl:maven-test:jar:1.0-SNAPSHOT':The following profiles are active: - profile1 (source: com.glowanet.appl:maven-test:1.0-SNAPSHOT)[INFO]
[INFO] --- echo-maven-plugin:1.3.2:echo (echo-plugin) @ maven-test ---
[INFO] plugin active
[INFO]
[INFO] --- echo-maven-plugin:1.3.2:echo (echo-profile) @ maven-test ---
[INFO] profile1 active
[INFO]
[INFO] --- maven-resources-plugin:2.6:resources (default-resources) @ maven-test ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] Copying 0 resource
[INFO]
[INFO] --- maven-compiler-plugin:3.1:compile (default-compile) @ maven-test ---
[INFO] Changes detected - recompiling the module!
[INFO] Compiling 1 source file to C:\projects\projects_github\maven-test\target\classes
[INFO]
[INFO] --- maven-resources-plugin:2.6:testResources (default-testResources) @ maven-test ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] skip non existing resourceDirectory C:\projects\projects_github\maven-test\src\test\resources
[INFO]
[INFO] --- maven-compiler-plugin:3.1:testCompile (default-testCompile) @ maven-test ---
[INFO] Nothing to compile - all classes are up to date
[INFO]
[INFO] --- maven-surefire-plugin:2.12.4:test (default-test) @ maven-test ---
[INFO] No tests to run.
[INFO]
[INFO] --- maven-jar-plugin:2.4:jar (default-jar) @ maven-test ---
[INFO] Building jar: C:\projects\projects_github\maven-test\target\maven-test-1.0-SNAPSHOT.jar
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  1.749 s
[INFO] Finished at: 2022-12-08T13:08:48+01:00
[INFO] ------------------------------------------------------------------------
```

### C O R R E C T !

# Finally

I'm able to confirm, that this fix will solve the problem.

Thx, Oliver

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

I will prepare a PR and run the ITs. Hope to get many eyes for the change.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Fixed with [828de7e1a8e44da991f113203400953302b8fed8](https://gitbox.apache.org/repos/asf?p=maven.git;a=commit;h=828de7e1a8e44da991f113203400953302b8fed8), with [3492f3ae997ef949d926201fc864c4a207fc7659](https://gitbox.apache.org/repos/asf?p=maven.git;a=commit;h=3492f3ae997ef949d926201fc864c4a207fc7659) for `maven-3.9.x` branch, with [006ea14902a5e807c2ef76c91b9d32ee3e8b4cfc](https://gitbox.apache.org/repos/asf?p=maven.git;a=commit;h=006ea14902a5e807c2ef76c91b9d32ee3e8b4cfc) for `maven-3.8.x` branch.
